File: script/R/06_rfecv.R
# ==============================================================================
# Step C — RFECV con LOGO-CV
# - Feature post correlation-pruning
# - Recursive Feature Elimination manuale con ranger + LOGO-CV
# - Trova numero ottimale di feature per balanced accuracy
# ==============================================================================
# Richiede: install.packages(c("ranger","ggplot2"))
# ==============================================================================

library(ranger)
library(ggplot2)

setwd(normalizePath("~/Desktop/IREN"))

RANDOM_SEED <- 42
N_TREES     <- 300
MIN_FEATS   <- 3

set.seed(RANDOM_SEED)

# ── Carica feature post-pruning ───────────────────────────────────────────────
lines_raw <- readLines("corr_pruned_features.txt")
feats     <- lines_raw[!grepl("^#", lines_raw) & nchar(trimws(lines_raw)) > 0]
cat(sprintf("Feature in input (post correlation pruning): %d\n", length(feats)))

# ── Carica training set ───────────────────────────────────────────────────────
df     <- read.csv("TRAIN_FEATURES.csv", sep=";", stringsAsFactors=FALSE)
groups <- df$Sample.ID
y      <- as.factor(df$Classe2)
X      <- df[, feats, drop=FALSE]
for (col in names(X))
    if (any(is.na(X[[col]]))) X[[col]][is.na(X[[col]])] <- median(X[[col]], na.rm=TRUE)

cat(sprintf("Training set: %d campioni, %d feature\n", nrow(df), length(feats)))

unique_groups <- unique(groups)
n_folds       <- length(unique_groups)
cat(sprintf("Gruppi LOGO: %d fold\n\n", n_folds))

# ── Helper: LOGO-CV con ranger su un sottoinsieme di feature ─────────────────
logo_balanced_accuracy <- function(feat_subset) {
    correct_per_class <- list()
    total_per_class   <- list()

    for (g in unique_groups) {
        test_mask  <- groups == g
        train_mask <- !test_mask

        X_tr <- X[train_mask, feat_subset, drop=FALSE]
        X_te <- X[test_mask,  feat_subset, drop=FALSE]
        y_tr <- y[train_mask]
        y_te <- y[test_mask]

        train_df_fold <- cbind(X_tr, Classe2=y_tr)
        rf <- ranger(Classe2 ~ ., data=train_df_fold,
                     num.trees=N_TREES, seed=RANDOM_SEED,
                     num.threads=parallel::detectCores())
        preds <- predict(rf, data=X_te)$predictions

        for (cls in levels(y)) {
            mask_true <- y_te == cls
            if (!sum(mask_true)) next
            correct_per_class[[cls]] <- (correct_per_class[[cls]] %||% 0) +
                sum(preds[mask_true] == cls)
            total_per_class[[cls]]   <- (total_per_class[[cls]] %||% 0) +
                sum(mask_true)
        }
    }
    per_class_acc <- sapply(levels(y), function(cls) {
        if (is.null(total_per_class[[cls]]) || total_per_class[[cls]] == 0) return(NA)
        correct_per_class[[cls]] / total_per_class[[cls]]
    })
    mean(per_class_acc, na.rm=TRUE)
}
`%||%` <- function(a, b) if (!is.null(a)) a else b

# ── RFE manuale ───────────────────────────────────────────────────────────────
# 1. Fit RF su tutte le feature → importanza
# 2. Evalua LOGO-CV per ogni sottoinsieme (da n_max a MIN_FEATS)
# 3. Trova n ottimale

cat("Avvio RFECV manuale...\n")
cat(sprintf("  Valuto %d configurazioni (da %d a %d feature)\n",
            length(feats) - MIN_FEATS + 1, length(feats), MIN_FEATS))

# Ranking iniziale tramite RF su tutti i dati di training
full_train <- cbind(X, Classe2=y)
rf_full <- ranger(Classe2 ~ ., data=full_train,
                  num.trees=N_TREES, importance="impurity",
                  seed=RANDOM_SEED, num.threads=parallel::detectCores())
imp_order <- order(rf_full$variable.importance, decreasing=TRUE)
feats_ranked <- feats[imp_order]   # feature ordinate per importanza

n_sizes  <- seq(length(feats), MIN_FEATS)
ba_scores <- numeric(length(n_sizes))

for (i in seq_along(n_sizes)) {
    n   <- n_sizes[i]
    top <- feats_ranked[1:n]
    ba  <- logo_balanced_accuracy(top)
    ba_scores[i] <- ba
    cat(sprintf("  n=%2d feature → balanced_accuracy=%.4f\n", n, ba))
}

best_idx   <- which.max(ba_scores)
n_optimal  <- n_sizes[best_idx]
ba_optimal <- ba_scores[best_idx]
selected_feats <- feats_ranked[1:n_optimal]

cat(sprintf("\nNumero ottimale di feature: %d\n", n_optimal))
cat(sprintf("Balanced accuracy ottimale: %.4f\n", ba_optimal))

# ── Report ────────────────────────────────────────────────────────────────────
cat(sprintf("\n%s\n  RFECV — Feature selezionate (%d)\n%s\n",
            strrep("=",50), n_optimal, strrep("=",50)))
for (i in seq_along(selected_feats))
    cat(sprintf("  %2d. %s\n", i, selected_feats[i]))

excluded <- setdiff(feats_ranked, selected_feats)
if (length(excluded))
    cat(sprintf("\n  Feature escluse (%d): %s\n", length(excluded), paste(excluded, collapse=", ")))

# ── Plot curva RFECV ──────────────────────────────────────────────────────────
curve_df <- data.frame(N=n_sizes, BA=ba_scores)
p <- ggplot(curve_df, aes(x=N, y=BA)) +
    geom_line(color="#4878CF", linewidth=1.2) +
    geom_point(color="#4878CF", size=3) +
    geom_vline(xintercept=n_optimal, linetype="dashed",
               color="#D65F5F", linewidth=1) +
    geom_hline(yintercept=ba_optimal, linetype="dotted",
               color="#D65F5F", linewidth=0.8, alpha=0.7) +
    annotate("label", x=n_optimal, y=ba_optimal - 0.04,
             label=sprintf("%d feat\nBA=%.3f", n_optimal, ba_optimal),
             color="#D65F5F", size=3.5, fill="white") +
    scale_x_continuous(breaks=n_sizes) +
    labs(title="RFECV — Selezione numero ottimale di feature\nELLONA/IREN",
         x="Numero di feature", y="Balanced accuracy (LOGO-CV)") +
    theme_minimal(base_size=11) +
    theme(panel.grid.minor=element_blank())

ggsave("rfecv_results.png", p, width=10, height=5, dpi=150)
cat("\n✓  rfecv_results.png\n")

# ── Salva ─────────────────────────────────────────────────────────────────────
write.table(curve_df, "rfecv_curve.csv", sep=";", row.names=FALSE, quote=FALSE)

con <- file("rfecv_selected_features.txt","w")
writeLines(sprintf("# RFECV — %d feature selezionate", n_optimal), con)
writeLines(sprintf("# Scoring: balanced_accuracy | CV: LOGO (gruppi=Sample.ID)"), con)
writeLines(sprintf("# Balanced accuracy ottimale: %.4f", ba_optimal), con)
writeLines("", con)
writeLines(selected_feats, con)
close(con)

cat("✓  rfecv_curve.csv\n")
cat("✓  rfecv_selected_features.txt\n")

File: script/Python/06_rfecv.py
"""
Step C — RFECV con LOGO-CV
- Carica le feature post correlation-pruning
- Recursive Feature Elimination con Cross-Validation
- CV: LeaveOneGroupOut (gruppi = Sample.ID)
- Base estimator: RandomForestClassifier
- Trova il numero ottimale di feature massimizzando balanced_accuracy
- Output: rfecv_results.png, rfecv_selected_features.txt, rfecv_report.csv
"""

import pandas as pd
import numpy as np
from sklearn.ensemble import RandomForestClassifier
from sklearn.feature_selection import RFECV
from sklearn.impute import SimpleImputer
from sklearn.model_selection import LeaveOneGroupOut
import matplotlib
matplotlib.use("Agg")
import matplotlib.pyplot as plt

# ── Parametri ──────────────────────────────────────────────────────────────────
TRAIN_FILE  = "TRAIN_FEATURES.csv"
INPUT_FEATS = "corr_pruned_features.txt"
OUT_FEATS   = "rfecv_selected_features.txt"
OUT_REPORT  = "rfecv_report.csv"
OUT_PLOT    = "rfecv_results.png"
RANDOM_SEED = 42
N_TREES     = 300
SCORING     = "balanced_accuracy"   # robusto con classi sbilanciate
MIN_FEATS   = 3
# ──────────────────────────────────────────────────────────────────────────────

# ── Carica feature post-pruning ───────────────────────────────────────────────
with open(INPUT_FEATS) as f:
    feats = [l.strip() for l in f if l.strip() and not l.startswith("#")]
print(f"Feature in input (post correlation pruning): {len(feats)}")

# ── Carica training set ───────────────────────────────────────────────────────
df     = pd.read_csv(TRAIN_FILE, sep=";")
groups = df["Sample.ID"].values
y      = df["Classe2"].values
X_raw  = df[feats].values

X = SimpleImputer(strategy="median").fit_transform(X_raw)
print(f"Training set: {X.shape[0]} campioni, {X.shape[1]} feature")
print(f"Gruppi LOGO: {len(np.unique(groups))} fold")

# ── RFECV ─────────────────────────────────────────────────────────────────────
rf = RandomForestClassifier(
    n_estimators=N_TREES,
    max_features="sqrt",
    class_weight="balanced",
    n_jobs=-1,
    random_state=RANDOM_SEED
)

logo = LeaveOneGroupOut()

selector = RFECV(
    estimator=rf,
    step=1,
    cv=logo,
    scoring=SCORING,
    min_features_to_select=MIN_FEATS,
    n_jobs=-1
)

print(f"\nAvvio RFECV (scoring={SCORING}, step=1)...\n")
selector.fit(X, y, groups=groups)

n_optimal = selector.n_features_
print(f"Numero ottimale di feature: {n_optimal}")
print(f"Balanced accuracy media al n_ottimale: "
      f"{selector.cv_results_['mean_test_score'][n_optimal-MIN_FEATS]:.4f}")

# ── Risultati per numero di feature ───────────────────────────────────────────
mean_scores = selector.cv_results_["mean_test_score"]
std_scores  = selector.cv_results_["std_test_score"]
n_range     = np.arange(MIN_FEATS, len(feats)+1)

# ── Feature selezionate ───────────────────────────────────────────────────────
selected_mask   = selector.support_
selected_feats  = [f for f, s in zip(feats, selected_mask) if s]
ranking         = selector.ranking_

report_df = pd.DataFrame({
    "Feature": feats,
    "Ranking": ranking,
    "Selected": selected_mask
}).sort_values(["Selected", "Ranking"], ascending=[False, True])

# ── Report console ─────────────────────────────────────────────────────────────
print(f"\n{'':=<55}")
print(f"  RFECV — Feature selezionate ({n_optimal})")
print(f"{'':=<55}")
for _, row in report_df[report_df["Selected"]].iterrows():
    print(f"  rank={int(row['Ranking'])}  {row['Feature']}")

print(f"\n  Feature escluse da RFECV ({(~selected_mask).sum()}):")
for _, row in report_df[~report_df["Selected"]].iterrows():
    print(f"  rank={int(row['Ranking'])}  {row['Feature']}")

# ── Plot ───────────────────────────────────────────────────────────────────────
fig, ax = plt.subplots(figsize=(10, 5))

ax.plot(n_range, mean_scores, "o-", color="#4878CF",
        linewidth=2, markersize=5, label="Balanced accuracy (media LOGO)")
ax.fill_between(n_range,
                mean_scores - std_scores,
                mean_scores + std_scores,
                alpha=0.18, color="#4878CF", label="± std")
ax.axvline(n_optimal, color="#D65F5F", linestyle="--", linewidth=1.5,
           label=f"Ottimale: {n_optimal} feature")
ax.axhline(mean_scores[n_optimal - MIN_FEATS], color="#D65F5F",
           linestyle=":", linewidth=1, alpha=0.7)

ax.set_xlabel("Numero di feature", fontsize=12)
ax.set_ylabel(f"Balanced accuracy (LOGO-CV)", fontsize=12)
ax.set_title("RFECV — Selezione numero ottimale di feature\nELLONA/IREN", fontsize=13)
ax.set_xticks(n_range)
ax.legend(fontsize=10)
ax.grid(alpha=0.3)

# Annota il punto ottimale
opt_score = mean_scores[n_optimal - MIN_FEATS]
ax.annotate(f"  {n_optimal} feat\n  BA={opt_score:.3f}",
            xy=(n_optimal, opt_score),
            xytext=(n_optimal + 0.8, opt_score - 0.04),
            fontsize=9, color="#D65F5F",
            arrowprops=dict(arrowstyle="->", color="#D65F5F", lw=1.2))

plt.tight_layout()
plt.savefig(OUT_PLOT, dpi=150)
plt.close()
print(f"\n✓  {OUT_PLOT}")

# ── Salva ─────────────────────────────────────────────────────────────────────
report_df.to_csv(OUT_REPORT, sep=";", index=False)

with open(OUT_FEATS, "w") as f:
    f.write(f"# RFECV — {n_optimal} feature selezionate\n")
    f.write(f"# Scoring: {SCORING} | CV: LOGO (gruppi=Sample.ID)\n")
    f.write(f"# Balanced accuracy ottimale: "
            f"{mean_scores[n_optimal-MIN_FEATS]:.4f} ± "
            f"{std_scores[n_optimal-MIN_FEATS]:.4f}\n\n")
    for feat in selected_feats:
        f.write(feat + "\n")

print(f"✓  {OUT_REPORT}")
print(f"✓  {OUT_FEATS}")

# Salva anche la curva completa
curve_df = pd.DataFrame({
    "N_features":    n_range,
    "MeanBA":        mean_scores,
    "StdBA":         std_scores
})
curve_df.to_csv("rfecv_curve.csv", sep=";", index=False)
print(f"✓  rfecv_curve.csv")
